In [1]:
from pathlib import Path
import pandas as pd

In [2]:
# Project directories
project_root = Path.cwd()
if not (project_root / "data" / "raw").exists():
    project_root = project_root.parent

raw_dir = project_root / "data" / "raw"

# Load CSV snapshots
df_oct = pd.read_csv(raw_dir / "Saudi foreign aid projects 25M10.csv")
df_dec = pd.read_csv(raw_dir / "Saudi foreign aid projects 25M12.csv")
df_mar = pd.read_csv(raw_dir / "Saudi foreign aid projects 26M03.csv")

# Display dataset dimensions
summary = pd.DataFrame({
    "Snapshot": ["October 2025", "December 2025", "March 2026"],
    "Rows": [len(df_oct), len(df_dec), len(df_mar)],
    "Columns": [df_oct.shape[1], df_dec.shape[1], df_mar.shape[1]]
})

display(summary)

,Snapshot,Rows,Columns
0,October 2025,8371,7
1,December 2025,8415,7
2,March 2026,8802,7


In [3]:
missing_summary = pd.DataFrame({
    "Snapshot": ["October 2025", "December 2025", "March 2026"],
    "Missing Values": [
        df_oct.isna().sum().sum(),
        df_dec.isna().sum().sum(),
        df_mar.isna().sum().sum()
    ]
})

display(missing_summary)

,Snapshot,Missing Values
0,October 2025,0
1,December 2025,0
2,March 2026,0


In [4]:
duplicates_summary = pd.DataFrame({
    "Snapshot": ["October 2025", "December 2025", "March 2026"],
    "Total Rows": [len(df_oct), len(df_dec), len(df_mar)],
    "Exact Duplicates": [
        df_oct.duplicated().sum(),
        df_dec.duplicated().sum(),
        df_mar.duplicated().sum()
    ],
    "Unique Rows": [
        len(df_oct.drop_duplicates()),
        len(df_dec.drop_duplicates()),
        len(df_mar.drop_duplicates())
    ]
})

display(duplicates_summary)

,Snapshot,Total Rows,Exact Duplicates,Unique Rows
0,October 2025,8371,671,7700
1,December 2025,8415,694,7721
2,March 2026,8802,801,8001


In [5]:
print("October 2025:", df_oct.columns.tolist())
print("December 2025:", df_dec.columns.tolist())
print("March 2026:", df_mar.columns.tolist())

October 2025: ['عنوان_المشروع', 'حالة_المشروع', 'قطاع_المشروع', 'موقع_المشروع', 'مجموع_التكاليف_بالدولار', 'الجهة_المانحة_للمشروع', 'الدولة_المستفيدة']
December 2025: ['عنوان المشروع', 'حالة المشروع', 'قطاع المشروع', 'موقع المشروع', 'مجموع التكاليف', 'الجهة المانحة للمشروع', 'الدولة المستفيدة']
March 2026: ['عنوان المشروع', 'حالة المشروع', 'قطاع المشروع', 'موقع المشروع', 'مجموع التكاليف $', 'الجهة المانحة للمشروع', 'الدولة المستفيدة']


In [6]:
unique_summary = pd.DataFrame({
    "Snapshot": ["October 2025", "December 2025", "March 2026"],
    "Countries": [
        df_oct.iloc[:, 6].nunique(),
        df_dec.iloc[:, 6].nunique(),
        df_mar.iloc[:, 6].nunique()
    ],
    "Sectors": [
        df_oct.iloc[:, 2].nunique(),
        df_dec.iloc[:, 2].nunique(),
        df_mar.iloc[:, 2].nunique()
    ],
    "Donors": [
        df_oct.iloc[:, 5].nunique(),
        df_dec.iloc[:, 5].nunique(),
        df_mar.iloc[:, 5].nunique()
    ],
    "Statuses": [
        df_oct.iloc[:, 1].nunique(),
        df_dec.iloc[:, 1].nunique(),
        df_mar.iloc[:, 1].nunique()
    ]
})

display(unique_summary)

,Snapshot,Countries,Sectors,Donors,Statuses
0,October 2025,174,36,21,2
1,December 2025,174,36,21,2
2,March 2026,176,36,22,2


In [7]:
snapshots = {
    "October 2025": df_oct,
    "December 2025": df_dec,
    "March 2026": df_mar
}

composite_results = []

for snapshot, df in snapshots.items():
    # First column: Project Title
    # Fifth column: Project Cost
    key_columns = [df.columns[0], df.columns[4]]

    repeated = df.duplicated(subset=key_columns, keep=False)

    composite_results.append({
        "Snapshot": snapshot,
        "Total Rows": len(df),
        "Unique Composite Keys": df.drop_duplicates(
            subset=key_columns
        ).shape[0],
        "Repeated Key Rows": repeated.sum(),
        "Extra Duplicate Rows": df.duplicated(
            subset=key_columns
        ).sum()
    })

composite_summary = pd.DataFrame(composite_results)

display(composite_summary)

,Snapshot,Total Rows,Unique Composite Keys,Repeated Key Rows,Extra Duplicate Rows
0,October 2025,8371,7549,1052,822
1,December 2025,8415,7570,1076,845
2,March 2026,8802,7842,1206,960


In [8]:
for snapshot, df in snapshots.items():
    print(f"\n{snapshot} — Data Types")

    display(pd.DataFrame({
        "Column": df.columns,
        "Data Type": df.dtypes.astype(str).values
    }))


October 2025 — Data Types


,Column,Data Type
0,عنوان_المشروع,object
1,حالة_المشروع,object
2,قطاع_المشروع,object
3,موقع_المشروع,object
4,مجموع_التكاليف_بالدولار,object
5,الجهة_المانحة_للمشروع,object
6,الدولة_المستفيدة,object



December 2025 — Data Types


,Column,Data Type
0,عنوان المشروع,object
1,حالة المشروع,object
2,قطاع المشروع,object
3,موقع المشروع,object
4,مجموع التكاليف,object
5,الجهة المانحة للمشروع,object
6,الدولة المستفيدة,object



March 2026 — Data Types


,Column,Data Type
0,عنوان المشروع,object
1,حالة المشروع,object
2,قطاع المشروع,object
3,موقع المشروع,object
4,مجموع التكاليف $,object
5,الجهة المانحة للمشروع,object
6,الدولة المستفيدة,object


In [9]:
for snapshot, df in snapshots.items():
    cost_column = df.columns[4]

    cost_numeric = pd.to_numeric(
        df[cost_column].astype(str).str.replace(",", "", regex=False).str.strip(),
        errors="coerce"
    )

    print(f"{snapshot}:")
    print(f"Valid numeric values: {cost_numeric.notna().sum()}")
    print(f"Invalid numeric values: {cost_numeric.isna().sum()}")
    print()

October 2025:
Valid numeric values: 8371
Invalid numeric values: 0

December 2025:
Valid numeric values: 8415
Invalid numeric values: 0

March 2026:
Valid numeric values: 8802
Invalid numeric values: 0



## Data Profiling Summary

Three snapshots of Saudi Foreign Aid Projects were analyzed:

- October 2025: 8,371 records
- December 2025: 8,415 records
- March 2026: 8,802 records

### Key Findings

- No missing values were detected.
- Exact duplicate records exist in all three snapshots.
- Potential duplicate records were identified using Project Title and Cost.
- Column names differ between snapshots and require standardization.
- All columns were initially loaded as object data types.
- All project cost values can be converted to numeric values after removing commas.

### Data Cleaning Recommendations

- Standardize column names across snapshots.
- Convert project cost values to numeric data types.
- Preserve original raw CSV files.
- Flag potential duplicates for further investigation.
- Keep snapshots separate to avoid double-counting projects across reporting periods.